### PR TITLE
use getConformingProtocols when printing opaque generic types

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
Resolves rdar://93610106

When printing a function with an opaque `some MyProtocol` type in its argument list from a serialized AST, the current mechanism to print that protocol fails. This PR updates that code to use a check that works on both source files and serialized ASTs.

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
@swift-ci Please smoke test

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -6387,10 +6387,27 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
...
       // Print based on the type.
       Printer << "some ";
-      if (auto inheritedType = decl->getInherited().front().getType())
-        inheritedType->print(Printer, Options);
-      else
-        Printer << "Any";
+      if (!decl->getConformingProtocols().empty()) {
+        ArrayRef<ProtocolDecl *> protocols = decl->getConformingProtocols();
+        bool printed = false;
+        for (unsigned i = 0, e = protocols.size(); i < e; i++) {
```
Comment (DougGregor) [inline lib/AST/ASTPrinter.cpp:6393]: You can use LLVM's `interleave` to do this well, and it'll handle when the print the separator and such.

File: lib/AST/ASTPrinter.cpp [hunk 2]
```diff
@@ -6387,10 +6387,27 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
...
+          if (auto printType = protocols[i]->getDeclaredType()) {
+            if (printed)
+              Printer << " & ";
+
+            printType->print(Printer, Options);
+
+            printed = true;
+          }
```
Comment (DougGregor) [inline lib/AST/ASTPrinter.cpp:6401]: If there's no declared type (which I think can't happen), you can print `protocols[i]->getName()` instead, so nothing will get dropped.

File: lib/AST/ASTPrinter.cpp (hunk 1 above)
Comment (QuietMisdreavus) [reply lib/AST/ASTPrinter.cpp]: Good call! I've pushed up a commit that switches to this.

**Comment by QuietMisdreavus** [comment]
@swift-ci Please smoke test

**Review by DougGregor — APPROVED** [review]

**Pull request merged**